README.md
# morgan1

TODO: Enter the cookbook description here.



CHANGELOG.md
# morgan1 CHANGELOG

This file is used to list changes made in each version of the morgan1 cookbook.

# 0.1.0

Initial release.

- change 0
- change 1



spec/unit/recipes/web_spec.rb
#
# Cookbook:: morgan1
# Spec:: web
#
# Copyright:: 2018, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'morgan1::web' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7.4.1708' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.4.1708')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


spec/unit/recipes/default_spec.rb
#
# Cookbook:: morgan1
# Spec:: default
#
# Copyright:: 2018, The Authors, All Rights Reserved.

require 'spec_helper'

describe 'morgan1::default' do
  context 'When all attributes are default, on Ubuntu 16.04' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'ubuntu', version: '16.04')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end

  context 'When all attributes are default, on CentOS 7.4.1708' do
    let(:chef_run) do
      # for a complete list of available platforms and versions see:
      # https://github.com/customink/fauxhai/blob/master/PLATFORMS.md
      runner = ChefSpec::ServerRunner.new(platform: 'centos', version: '7.4.1708')
      runner.converge(described_recipe)
    end

    it 'converges successfully' do
      expect { chef_run }.to_not raise_error
    end
  end
end


test/integration/default/web_test.rb
# # encoding: utf-8

# Inspec test for recipe morgan1::web

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

unless os.windows?
  # This is an example test, replace with your own test.
  describe user('root'), :skip do
    it { should exist }
  end
end

# This is an example test, replace it with your own test.
describe port(80), :skip do
  it { should_not be_listening }
end


recipes/web.rb
#
# Cookbook:: morgan1
# Recipe:: web
#
# Copyright:: 2018, The Authors, All Rights Reserved.

powershell_script 'install Nuget package provider' do
    code 'Install-PackageProvider -Name NuGet -Force'
    not_if '(Get-PackageProvider -Name Nuget -ListAvailable -ErrorAction SilentlyContinue) -ne $null'
end
  
powershell_package_source 'Nuget' do
    action :register
    provider_name 'NuGet'
    url 'https://www.nuget.org/api/v2/'
    trusted true

end

powershell_package 'Install xWebAdministration' do
    action :install
    package_name 'xWebAdministration'
    version '2.3.0.0'
end

powershell_package 'Install xPSDesiredStateConfiguration' do
    action :install
    package_name 'xPSDesiredStateConfiguration'
    version '8.4.0.0'
end

powershell_package 'Install NetworkingDsc' do
    action :install
    package_name 'NetworkingDsc'
    version '6.1.0.0'
end


dsc_resource "stop default" do
    resource :xWebsite
    property :name, 'Default Web Site'
    property :ensure, 'Present'
    property :state, 'Stopped'
end

dsc_resource "setup-dir-1" do
    resource :file
    property :ensure, 'Present'
    property :type, 'Directory'
    property :destinationpath, 'C:\Packages'
end

dsc_resource "setup-dir-2" do
    resource :file
    property :ensure, 'Present'
    property :type, 'Directory'
    property :destinationpath, 'C:\app'
end

dsc_resource "SchUseStrongCrypto" do
    resource :registry
    property :ensure, 'Present'
    property :key, 'HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\.NETFramework\v4.0.30319'
    property :valuename, 'SchUseStrongCrypto'
    property :valuetype, 'Dword'
    property :valuedata, ['1']
end

dsc_resource "SchUseStrongCrypto64" do
    resource :registry
    property :ensure, 'Present'
    property :key, 'HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Microsoft\.NETFramework\v4.0.30319'
    property :valuename, 'SchUseStrongCrypto'
    property :valuetype, 'Dword'
    property :valuedata, ['1']
end

dsc_resource "download web site" do
    resource :xremotefile
    property :ensure, 'Present'
    property :uri, 'http://github.com/MorganPeat/Pilot/releases/download/pilot-v1.0.15/Pilot.WebApi.zip'
    property :destinationpath, 'C:\packages\Pilot.WebApi.zip'
end

dsc_resource "unzip web site" do
    resource :archive
    property :ensure, 'Present'
    property :path, 'c:\packages\Pilot.WebApi.zip'
    property :destination, 'C:\app'
end


dsc_resource "app pool" do
    resource :xWebAppPool
    property :ensure, 'Present'
    property :name, 'Pilot.WebApi.AppPool'
    property :state, 'Started'
    property :autoStart, true
    property :enable32BitAppOnWin64, true
    property :managedPipelineMode, 'Integrated'
    property :managedRuntimeVersion, 'v4.0'
end

directory "c:\\app" do
    rights :read, 'IIS_IUSRS'
    recursive true
    action :create
end

# https://www.dotnetcatch.com/2016/05/11/dsc-in-chef-converting-ruby-hashes-to-powershell-types/
class WebsiteBindings 
    @bindings = [] 
    def initialize(bindings) 
      @bindings = bindings 
    end
       
    def to_psobject() 
      bindings = Array.new() 
      @bindings.each do |b| 
        bindings.push("(new-ciminstance -classname MSFT_xWebBindingInformation -Namespace root/microsoft/Windows/DesiredStateConfiguration -Property @{Protocol='#{b[:protocol]}';IPAddress='#{b[:ip]}';Port=#{b[:port]}} -ClientOnly)") 
      end
      "[ciminstance[]](#{bindings.join(',')})"
    end
  end

node.run_state['bindings'] = WebsiteBindings.new([ 
{ protocol: 'http', ip: '*', port: 8080 }, 
{ protocol: 'http', ip: '*', port: 8081 } ]) 


dsc_resource "create website" do
    resource :xWebsite
    property :ensure, 'Present'
    property :name, 'Pilot.WebApi'
    property :physicalPath, 'C:\app'
    property :state, 'Started'
    property :bindingInfo, node.run_state['bindings']
    property :applicationPool, 'Pilot.WebApi.AppPool'
end

dsc_resource "firewall rule" do
    resource :Firewall
    property :ensure, 'Present'
    property :name, 'PilotWebApiRule'
    property :displayName, 'Pilot.WebApi (TCP 8080 in)'
    property :action, 'Allow'
    property :direction, 'Inbound'
    property :localPort, ['8080']
    property :protocol, 'TCP'
    property :profile, ['Any']
    property :enabled, 'True'
end

recipes/add_iis_application.rb
#
# Cookbook:: mogran1
# Recipe:: add_iis_application
#
# Copyright:: 2018, NWM, All Rights Reserved.

# Grant IIS (w3wp.exe) rights to read the binaries
directory node['iis']['physicalPath'] do
  rights :read, 'IIS_IUSRS'
  recursive true
  action :create
end

# https://www.dotnetcatch.com/2016/05/11/dsc-in-chef-converting-ruby-hashes-to-powershell-types/
class WebsiteBindings
  @bindings = []
  def initialize(bindings)
    @bindings = bindings
  end

  def to_psobject
    bindings = []
    @bindings.each do |b|
      bindings.push("(new-ciminstance -classname MSFT_xWebBindingInformation -Namespace root/microsoft/Windows/DesiredStateConfiguration -Property @{Protocol='#{b[:protocol]}';IPAddress='#{b[:ip]}';Port=#{b[:port]}} -ClientOnly)")
    end
    "[ciminstance[]](#{bindings.join(',')})"
  end
end

dsc_resource 'Create App Pool' do
  resource :xWebAppPool
  property :name, node['iis']['name']
  property :state, node['iis']['pool-state']
  property :autoStart, true
  property :enable32BitAppOnWin64, node['iis']['32-bit-pool']
  property :managedPipelineMode, 'Integrated'
  property :managedRuntimeVersion, 'v4.0'
end

dsc_resource 'Create Website' do
  resource :xWebsite
  property :name, node['iis']['name']
  property :physicalPath, node['iis']['physicalPath']
  property :state, node['iis']['site-state']
  property :bindingInfo, WebsiteBindings.new(node['iis']['bindings'])
  property :applicationPool, node['iis']['name']
end

# Open a port for each IIS binding
node['iis']['bindings'].each do |binding|
  dsc_resource "Open Firewall (#{binding[:port]})" do
    resource :Firewall
    property :name, "#{node['iis']['name']}-#{binding[:port]}"
    property :displayName, "#{node['iis']['name']} (TCP #{binding[:port]} in)"
    property :action, 'Allow'
    property :direction, 'Inbound'
    property :localPort, [binding[:port].to_s]
    property :protocol, 'TCP'
    property :profile, ['Any']
    property :enabled, 'True'
  end
end


recipes/setup_prerequisites.rb
# Cookbook:: morgan1
# Recipe:: setup_prerequisites
#
# Copyright:: 2018, NWM, All Rights Reserved.

# Ensure all necessary windows features are installed
node['iis']['windows-features'].each do |feature|
  dsc_resource "Install #{feature}" do
    resource :WindowsFeature
    property :Name, feature
  end
end

# Sets up the source of powershell packages
# TODO: Artifactory?
powershell_script 'Install-PackageProvider (NuGet)' do
  code 'Install-PackageProvider -Name NuGet -Force'
  not_if '(Get-PackageProvider -Name Nuget -ListAvailable -ErrorAction SilentlyContinue) -ne $null'
end

powershell_package_source 'Nuget' do
  action :register
  provider_name 'NuGet'
  url 'https://www.nuget.org/api/v2/'
  trusted true
end

node['iis']['powershell-packages'].each do |package, version|
  powershell_package "Install package #{package}" do
    action :install
    package_name package
    version version
  end
end

# Use strong crypto (>= TLS 1.1) by default
# https://docs.microsoft.com/en-us/dotnet/framework/network-programming/tls#configuring-security-via-the-windows-registry
[
  'HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\.NETFramework\v4.0.30319',
  'HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Microsoft\.NETFramework\v4.0.30319',
].each do |key|
  dsc_resource 'SchUseStrongCrypto' do
    resource :registry
    property :key, key
    property :valuename, 'SchUseStrongCrypto'
    property :valuetype, 'Dword'
    property :valuedata, ['1']
  end
end

# Ensure the correct .NET Framework is installed
dsc_resource 'Download .NET Framework installer' do
  resource :xRemoteFile
  property :Uri, node['dotnet']['uri']
  property :DestinationPath, "#{Chef::Config[:file_cache_path]}\\dotnet-installer.exe"
end

reboot 'Restart Computer' do
  action :nothing
  reason 'Reboot after .NET Framework install'
end

dsc_resource 'Install .NET Framework' do
  resource :cDotNetFramework
  property :KB, node['dotnet']['kb']
  property :InstallerPath, "#{Chef::Config[:file_cache_path]}\\dotnet-installer.exe"
  property :Ensure, 'Present'
  notifies :reboot_now, 'reboot[Restart Computer]', :immediately
end


attributes/default.rb
# frozen_string_literal: true

# Cookbook Name::morgan1
# Attributes::default.rb

###########################################################################################
# TODOs / Questions
#
# - Is the attributes file the best place to list features, DSC packages? Don't really want
#   to give callers the ability to modify / change it, it should be a bit more private to
#   the recipe.
# - Is there a way of forcing an attribute to be supplied? e.g. the artifacts url must be
#   supplied, there is no suitable default value, and the script should error with a suitable
#   message if a value is missing.
# - Where should we pick up Powershell DSC packages from? Artifactory? Can we get it to
#   act like a mirror, c.f. nuget?
###########################################################################################

# Windows features required for IIS applications to run
# TODO: CTOM-271 Determine correct Windows feature list for IIS deployment
# Use Get-WindowsFeature to see available features
default['iis']['windows-features'] = [
  'Web-Server',
  'Web-Common-Http',
  'Web-Http-Errors',
  'Web-Static-Content',
  'Web-Health',
  'Web-Http-logging',
  'Web-Log-Libraries',
  'Web-Performance',
  'Web-Stat-Compression',
  'Web-Dyn-Compression',
  'Web-Security',
  'Web-Filtering',
  'Web-Client-Auth',
  'Web-Digest-Auth',
  'Web-Cert-Auth',
  'Web-IP-Security',
  'Web-Url-Auth',
  'Web-Windows-Auth',
  'Web-Net-Ext45',
  'Web-Asp-Net45',
  'Web-WebSockets',
  'Web-Mgmt-Tools',
]

# Powershell packges required during Chef bootstrapping
default['iis']['powershell-packages'] = {
  'xWebAdministration'           => '2.3.0.0',
  'xPSDesiredStateConfiguration' => '8.4.0.0',
  'NetworkingDsc'                => '6.1.0.0',
  'GraniResource'                => '3.7.11.0',
}

# Name of the IIS website
default['iis']['name'] = ''
# URI where artifact is downloaded from (zip file, extracted to site physical path)
default['iis']['artifact_uri'] = ''
# Path where artifact is extracted to; physical path for IIS web site
default['iis']['physicalPath'] = 'C:\inetpub\wwwroot'
# Web site binding(s); an ingress firewall rule is created for each port
default['iis']['bindings'] = [
  { protocol: 'http', ip: '*', port: 80 },
  { protocol: 'https', ip: '*', port: 443 }
]
# Site is started or stopped on creation
default['iis']['site-state'] = 'Started'
# App pool is started or stopped on creation
default['iis']['pool-state'] = 'Started'
# 32- or 64-bit app pool process
default['iis']['32-bit-pool'] = false

# Values taken from https://support.microsoft.com/en-gb/help/4054531/microsoft-net-framework-4-7-2-web-installer-for-windows
default['dotnet']['uri'] = 'http://go.microsoft.com/fwlink/?LinkId=863262'
default['dotnet']['kb'] = 'KB4054590'
